File: ctci/tmp/chapter_4_trees_graphs/c/4.8-first_common_ancestor.c
#include "tree.h"

/* LINK TO PARENTS EXIST */

/**
 * Finds the depth from a node of a binary tree. 
 */
unsigned int find_depth(tree_node *node)
{
    unsigned int depth = 0;

    while (node)
    {
        node = node->parent;
        depth++;
    }
    return depth;
}

/**
 * Moves a node up by a given number of nodes. 
 */
tree_node *move_up_by(tree_node *node, unsigned int delta)
{
    tree_node *curr;

    curr = node;
    while (delta && curr)
    {
        curr = curr->parent;
        delta--;
    }
    return curr;
}

/* Finds the first common ancestor of two nodes in a binary tree if there are
 * links to parents. */
tree_node *find_common_ancestor(tree_node *p, tree_node *q)
{
    int delta;
    tree_node *first, *second;

    delta = find_depth(p) - find_depth(q);
    first = delta > 0 ? p : q;
    second = delta > 0 ? q : p;
    first = move_up_by(first, abs(delta));
    while (first && second && first != second)
    {
        first = first->parent;
        second = second->parent;
    }
    return first && second ? first : NULL;
}

/* LINK TO PARENTS DOES NOT EXIST */

/**
 * Determines if a node has another node in subtree.
 */
int covers(tree_node *root, tree_node *node)
{
    if (!root)
        return 0;
    if (root == node)
        return 1;
    return covers(root->left, node) || covers(root->right, node);
}

/**
 * Helper.
 */
tree_node *common_ancestor_helper(tree_node *root, tree_node *p, tree_node *q)
{
    int is_p_left;
    int is_q_left;
    tree_node *child_side;

    if (!root || root == p || root == q)
        return root;
    is_p_left = covers(root->left, p);
    is_q_left = covers(root->left, q);
    if (is_p_left != is_q_left)
        return root;
    child_side = is_p_left ? root->left : root->right;
    return common_ancestor_helper(child_side, p, q);
}

/**
 * Finds the first common ancestor of two nodes in a binary tree if there
 * are no links to parents. */
tree_node *common_ancestor(tree_node *root, tree_node *p, tree_node *q)
{
    if (!covers(root, p) || !covers(root, q))
        return NULL;
    return common_ancestor_helper(root, p, q);
}

int main(void)
{
    tree_node *root = create_node(1);
    tree_node *n2 = add_left(root, 2);
    tree_node *n3 = add_right(root, 3);
    tree_node *n4 = add_left(n2, 4);
    tree_node *n5 = add_right(n2, 5);
    tree_node *n6 = add_left(n3, 6);
    tree_node *n7 = add_right(n3, 7);
    tree_node *n8 = add_left(n4, 8);
    tree_node *n9 = add_right(n4, 9);
    tree_node *n10 = add_left(n5, 10);
    tree_node *n11 = add_right(n5, 11);

    printf("%i\n", find_common_ancestor(n8, n11)->value);  // 2
    printf("%i\n", find_common_ancestor(n8, n7)->value);   // 1
    printf("%i\n", find_common_ancestor(n6, n7)->value);   // 3
    printf("%i\n", common_ancestor(root, n8, n11)->value); // 2
    printf("%i\n", common_ancestor(root, n8, n7)->value);  // 1
    printf("%i\n", common_ancestor(root, n6, n7)->value);  // 3

    return 0;
}

File: ctci/chapter_4_trees_graphs/4.8-first_common_ancestor.py
TreeNode = __import__('tree').TreeNode

# LINK TO PARENTS EXIST


def first_common_ancestor(p: TreeNode, q: TreeNode) -> TreeNode:
    """
    Finds the first common ancestor of two nodes in a binary tree
    if there are links to parents.
    """
    delta = find_depth(p) - find_depth(q)
    if delta > 0:
        first = p
        second = q
    else:
        first = q
        second = p
    first = move_up_by(first, abs(delta))
    while first and second and second != first:
        second = second.parent
        first = first.parent
    if first and second:
        return first
    return None


def move_up_by(node: TreeNode, delta: int) -> TreeNode:
    """
    Moves a node up by a given number of nodes.
    """
    curr = node
    while delta > 0 and curr:
        curr = curr.parent
        delta -= 1
    return curr


def find_depth(node: TreeNode) -> int:
    """
    Finds the depth from a node of a binary tree.
    """
    depth = 0
    while node:
        node = node.parent
        depth += 1
    return depth

# LINK TO PARENTS DOES NOT EXIST


def common_ancestor(root: TreeNode, first: TreeNode, second: TreeNode) -> TreeNode:
    """
    Finds the first common ancestor of two nodes in a binary tree
    if there are no links to parents.
    """
    # Error check - if one node is not in the tree.
    if not covers(root, first) or not covers(root, second):
        return None
    return common_ancestor_helper(root, first, second)


def common_ancestor_helper(root: TreeNode, p: TreeNode, q: TreeNode) -> TreeNode:
    """
    Helper.
    """
    if not root or p == q or q == root:
        return root
    first_on_left = covers(root.left, p)
    second_on_left = covers(root.left, q)
    if first_on_left != second_on_left:
        return root
    child_side = root.left if first_on_left else root.right
    return common_ancestor_helper(child_side, p, q)


def covers(root: TreeNode, node: TreeNode) -> bool:
    """
    Determines if a node has another node in subtree.
    """
    if not root:
        return False
    if root == node:
        return True
    return covers(root.left, node) or covers(root.right, node)


if __name__ == "main":
    n1 = TreeNode(1)
    n2 = n1.add_left(2)
    n3 = n1.add_right(3)
    n4 = n2.add_left(4)
    n5 = n2.add_right(5)
    n6 = n3.add_left(6)
    n7 = n3.add_right(7)
    n8 = n4.add_left(8)
    n9 = n4.add_right(9)
    n10 = n5.add_left(10)
    n11 = n5.add_right(11)
    n12 = n6.add_left(12)
    n13 = n6.add_right(13)
    n15 = TreeNode(15)

    print(first_common_ancestor(n9, n5))  # n2
    print(first_common_ancestor(n9, n15))  # None

    print(common_ancestor(n1, n9, n5))  # n2
    print(common_ancestor(n1, n9, n15))  # None
